File: 23.py
#Project Euler solution 23, author: http://github.com/TerkaSlan #
from time import clock
import math

def is_abundant(num):
    if num < 12:
        return False
    sum = 1
    for x in range(2,(int)(math.sqrt(num))+1):
        if num%x == 0:
            sum += x + num/x
    # correct is the sum is a perfect square
    if math.sqrt(num) == (int)(math.sqrt(num)):
        sum -= math.sqrt(num)
    return sum>num

start = clock()
abundants = list(i for i in range(1,28123) if is_abundant(i))

total_sum = 0
is_rewritable_as_sum = False

# guarantees there are no duplicated in abundants
abundants = set(abundants)
# according to wolphram mathematica: Every num > 20161 can be expressed as sum of 2 abundants
for i in range(1,20161+1,1):
    # any() returns true if at least 1 element is in abundants
    if not any( (i-abundant in abundants) for abundant in abundants):
        total_sum += i

end = clock()
print total_sum
print 'Took me ', end-start, ' seconds to come up with a solution' # 2.103s

File: 23.c
//Project Euler solution 23, author: http://github.com/TerkaSlan #
#include <stdio.h>
#include <string.h>
#include <time.h>
#define LIMIT 28123
unsigned sum_of_divisors(unsigned n){
	unsigned sum = 0;
	for(int i=1; i*i<=n; i++){
		if(n%i==0){
			sum+=i;
			// using found divisor i can
			// divide n by it to find the other one
			if(i != n/i && n != n/i)
				sum+=n/i;
		}
	}
	return sum;
}

int main(void)
{
	clock_t start, end;
	start = clock();
	unsigned count=0, num_of_abundants=0, abundant_nums[LIMIT];
	for(int n=1; n<LIMIT+1; n++){
		if(sum_of_divisors(n) > n){
			abundant_nums[num_of_abundants++] = n;
		}
	}

	long unsigned total_sum = 0;
	// by changing value at index i I mark the numbers not to count
	int all_nums[LIMIT];
	memset(all_nums, '\0', LIMIT);

	for(int i=0; i<num_of_abundants; i++){
		count = i;
		while(abundant_nums[i] + abundant_nums[count] < LIMIT){
			all_nums[abundant_nums[i] + abundant_nums[count]] = 1;
			count++;
		}
	}
	for (int j = 0; j < LIMIT; j++)
	{
		if(all_nums[j] != 1)
			total_sum += j;
	}
	end = clock();
	printf("Found: %lu \n", total_sum);
	printf("Took me %f seconds to come up with a solution\n", ((end-start)/CLOCKS_PER_SEC)*1.0); // 0s :D		
	return 0;
}